Add new API reference style and extension plugin guide (#743)
asreview/asreview: docs/source/extensions/extension_development.rst

    │       ├── balance
    │       │   ├── __init__.py
    │       │   └── example_balance_strategies.py
    │       └── query
    │           ├── __init__.py
    │           └── example_query_strategies.py
    ├── setup.py
    └── tests

The next step is to add metadata to the `setup.py
<https://github.com/asreview/template-extension-new-model/blob/main/setup.py>`__
file. Edit the ``name`` of the package and point the ``entry_points`` to the
models.

.. code:: bash

    entry_points={
        'asreview.models.classifiers': [
            'example = asreviewcontrib.models.classifiers.example_model:ExampleClassifier',
        ],
        'asreview.models.feature_extraction': [
            # define feature_extraction algorithms
        ],
        'asreview.models.balance': [
            # define balance_strategy algorithms
        ],
        'asreview.models.query': [
            # define query_strategy algorithms
        ]
    },

This code registers the model with name ``example``.

.. |template_link| raw:: html

    <a href="https://github.com/asreview/template-extension-new-model"
    target="_blank"> template for extending ASReview</a>

> 
> 
> 
> 
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```

Please remove the additional linebreaks.

any directory. The argument ``argv`` are the command line arguments left after
removing asreview and the entry point.

It is advised to place the newly defined class ``ExampleEntryPoints`` in the
following package structure:
``asreviewcontrib.{extension_name}.{your_modules}``. For example:

.. code:: bash

    ├── README.md
    ├── asreviewcontrib
    │   └── example
    │       ├── __init__.py
    │       ├── entrypoint.py
    │       └── example_utils.py
    ├── setup.py
    └── tests

Create a ``setup.py`` in
the root of the package, and set the keyword argument `entry_points` of
``setup()`` under ``asreview.entry_points``, for example:

.. code:: python

    entry_points={
        "asreview.entry_points": [
            "example = asreviewcontrib.example.entrypoint:ExampleEntryPoint",
        ]
    }

After installing this package, ASReview is extended with the ``asreview
example`` subcommand. See ``asreview -h`` for this option.

An example of a subcommand extension is the `Visualization Extension
<https://github.com/asreview/asreview-visualization>`_

Dataset extensions
------------------
An extension of the :class:`asreview.datasets.BaseDataSet` class.

> Dataset extensions integrate new datasets for use in ASReview. Adding datasets
> via extension provides quick access to the dataset via commandline or other user
> interface.

Reviewer: ```suggestion
Dataset extensions integrate new datasets for use in ASReview. Adding datasets
via extension provides quick access to the dataset via Command Line Interface or in
ASReview LAB.
```

Let's add cross-references to the relevant parts of CLI and LAB.

It is advised to place the new dataset ``your_dataset`` in the
following package structure:

.. code:: bash

    ├── README.md
    ├── asreviewcontrib
    │   └── dataset_name
    │       ├── __init__.py
    │       └── your_dataset.py
    ├── config

>     │   └── your_dataset.json

Reviewer: I would suggest not use the config option for this example. This is pretty advanced usage. This was used in ASReview-covid to make automated updates available. Most use cases don't need this I think.

>     ├── datasets
>     │   └── dataset_name

Reviewer: ```suggestion
    ├── data
```

>     │       ├── __init__.py

Reviewer: No need to have an `__init__.py` file in a data folder.

    │       └── your_dataset.csv
    ├── setup.py
    └── tests

> 

Reviewer: I think we need an example of `your_dataset.py` here. Covid plugin is relatively complicated.

I think a minimal example would look like this:

```
from asreview.datasets import BaseDataSet
from asreview.datasets import BaseDataGroup

class YourDataGroup(BaseDataGroup):
    group_id = "your_data_group"
    description = "A new data group with my awesome datasets."

    def __init__(self):
        
        dataset = BaseDataSet.from_config({
            "dataset_id": "your_data_id",
            "url": "",
            "reference": "",
            "link": "",
            "license": "",
            "title": "Your Data",
            "authors": [
              "Jane Doe",
              "John Doe"
            ],
            "year": 2021,
            "topic": "Your topic",
            "final_inclusions": true,
            "title_abstract_inclusions": false
          }
        )

        super(YourDataGroup, self).__init__(dataset)
        # pass multiple datasets to init if there are more datasets
```

For minimal functionality, ``your_dataset.py`` should extent

> :class:`asreview.datasets.BaseDataSet`. Further functionality can be extensions

Reviewer: Maybe also explain what `asreview.datasets.BaseDataGroup` does?

> of any other class in :mod:`asreview.datasets`. 

Reviewer: I would suggest creating a template repo here. Like https://github.com/asreview/template-extension-new-model

This should be straightforward with the example above.
